src/DataStore.ts

```diff
@@ -63,3 +64,91 @@ export function untrackVoiceConnection(guildId: string) {
 export function trackVoiceConnection(guildId: string, voiceConnection: VoiceConnection) {
 	return voiceConnections.set(guildId, voiceConnection);
 }
+
+// Audio Players
+
+// Each audio packet is 20ms long
+const FRAME_LENGTH = 20;
+
+let audioCycleInterval: NodeJS.Timeout | undefined;
+let nextTime = -1;
+
+/**
+ * A list of created audio players that are still active and haven't been destroyed.
+ */
+const audioPlayers: AudioPlayer[] = [];
+
+/**
+ * Called roughly every 20 milliseconds. Dispatches audio from all players, and then gets the players to prepare
+ * the next audio frame.
+ */
+function audioCycleStep() {
+	nextTime += FRAME_LENGTH;
+	const available = audioPlayers.filter((player) => player.checkPlayable());
+
+	// eslint-disable-next-line @typescript-eslint/dot-notation
+	available.forEach((player) => player['_stepDispatch']());
+
+	prepareNextAudioFrame(available);
+}
+
+/**
+ * Recursively gets the players that have been passed as parameters to prepare audio frames that can be played.
+ * at the start of the next cycle.
+ */
+function prepareNextAudioFrame(players: AudioPlayer[]) {
+	const nextPlayer = players.shift();
+
+	if (!nextPlayer) {
+		if (nextTime !== -1) {
+			audioCycleInterval = setTimeout(() => audioCycleStep(), nextTime - Date.now());
+		}
+		return;
+	}
+
+	// eslint-disable-next-line @typescript-eslint/dot-notation
+	nextPlayer['_stepPrepare']();
```

Comment:
What is this for?

Reply:
`_stepPrepare` is a private method so that it isn't exposed in the documentation. To call it without typescript complaining, I use this syntax.

Reply:
oh, okay